pontoon/base/tests/models/test_stats.py

```diff
@@ -0,0 +1,162 @@
+"""
+Test consistency of calculations between `calculate_stats` and `translation.save()`.
+"""
+import pytest
+from pontoon.base.models import TranslatedResource
+from pontoon.checks.models import (
+    Error,
+    Warning,
+)
+
+
+@pytest.fixture
+def translation_with_error(translation_a):
+    Error.objects.create(
+        translation=translation_a,
+        library='p',
+        message='error'
+    )
+    return translation_a
+
+
+@pytest.fixture
+def translation_with_warning(translation_a):
+    Warning.objects.create(
+        translation=translation_a,
+        library='p',
+        message='warning',
+    )
+    return translation_a
+
+
+def recalculate_stats(translation):
+    """
+    Make the full recalculate stats on a TranslatedResource.
+    """
+    translation.save(update_stats=False)
+    translated_resource = TranslatedResource.objects.get(
+        resource=translation.entity.resource,
+        locale=translation.locale,
+    )
+    translated_resource.calculate_stats()
+
+
+def diff_stats(t):
+    """
+    Update only necessary stats by calculating difference between stats.
+    """
+    t.save()
+
+
+@pytest.fixture(
+    params=(
+        recalculate_stats,
+        diff_stats,
+    )
+)
+def stats_update(db, request):
+    """
+    Wrapper fixture which allows to test both implementations of stats calculations.
+    """
+    return request.param
+
+
+@pytest.fixture
+def get_stats():
+    """
+    Fixture wrapper for function to get_stats of a TranslatedResource assigned to a translation.
+    """
+    def f(translation):
+        return TranslatedResource.objects.stats(
+            translation.entity.resource.project,
+            None,
+            translation.locale,
+        )
+    return f
+
+
+@pytest.fixture
+def stats_dict():
+    def f(**override_stats):
+        stats = {
+            'approved': 0,
+            'fuzzy': 0,
+            'total': 0,
+            'unreviewed': 0,
+            'warnings': 0,
+            'errors': 0,
+        }
+        stats.update(override_stats)
+        return stats
+    return f
```

Comment:
I'd rather be explicit and compare the output of `get_stats()` with the entire dict, otherwise it takes additional effort to figure out what's going on here. I think it's OK to repeat code in tests.

But that's just my preference, you can leave this unchanged.

Reply:
Readability > DRY :+1: